spikeinterface/toolkit/qualitymetrics/pca_metrics.py

```diff
@@ -280,31 +287,55 @@ def nearest_neighbors_isolation(all_pcs, all_labels, this_unit_id:int,
-
-        if spikes_for_nn_actual < n_spikes_other_unit:
-            pcs_other_unit_idx = rng.choice(np.arange(n_spikes_other_unit), size=spikes_for_nn_actual)
-            pcs_other_unit = pcs_other_unit[pcs_other_unit_idx]
-
-        isolation[other_unit_id==other_units_ids] = _compute_isolation(pcs_target_unit, pcs_other_unit, n_neighbors)
     
-    nn_isolation = np.min(isolation)
+    # get waveforms for target cluster
+    waveforms_target_unit = waveform_extractor.get_waveforms(unit_id=this_unit_id)
+    n_spikes_target_unit = waveforms_target_unit.shape[0]
+    
+    # find units whose closest channels overlap with closest channels of target cluster
+    closest_chans_all = get_template_channel_sparsity(waveform_extractor, method='radius',
+                                                      outputs='index', peak_sign='both', 
+                                                      radius_um=radius_um)
+    closest_chans_target_unit = closest_chans_all[this_unit_id]
+    other_units_ids = [unit_id for unit_id in other_units_ids if np.any(np.in1d(closest_chans_all[unit_id],
+                                                                                closest_chans_target_unit))]
+    
+    # if no unit is within neighborhood of target unit, then just say isolation is 1 (best possible)
+    if not other_units_ids:
+        nn_isolation = 1
+    # if there are units to compare, then compute isolation with each 
+    else: 
+        isolation = np.zeros(len(other_units_ids),)
+        for other_unit_id in other_units_ids:
+            waveforms_other_unit = waveform_extractor.get_waveforms(unit_id=other_unit_id)
+            n_spikes_other_unit = waveforms_other_unit.shape[0]
+
+            n_snippets = np.min([n_spikes_target_unit, n_spikes_other_unit, max_spikes_for_nn])
+
+            # make the two clusters equal in terms of:
+            # - number of spikes
+            # - channels with signal
+            waveforms_target_unit_idx = rng.choice(n_spikes_target_unit, size=n_snippets, replace=False)
+            waveforms_target_unit_sampled = waveforms_target_unit[waveforms_target_unit_idx]
+            waveforms_target_unit_sampled = waveforms_target_unit_sampled[:, :, closest_chans_target_unit]
+
+            waveforms_other_unit_idx = rng.choice(n_spikes_other_unit, size=n_snippets, replace=False)
+            waveforms_other_unit_sampled = waveforms_other_unit[waveforms_other_unit_idx]
+            waveforms_other_unit_sampled = waveforms_other_unit_sampled[:, :, closest_chans_target_unit]
+
+            # compute principal components after concatenation
+            all_snippets = np.concatenate([waveforms_target_unit_sampled.reshape((n_snippets, -1)),
+                                           waveforms_other_unit_sampled.reshape((n_snippets, -1))], axis=0)
+            pca = IncrementalPCA(n_components=n_components)
+            pca.partial_fit(all_snippets)
+            projected_snippets = pca.transform(all_snippets)
```

Comment:
@khl02007 since the pca is already computed, is there a reason to re-fit it? Is it to get the concatenated version?

Reply:
@alejoe91 Right, the PCs are computed for the concatenated waveforms of the pair of clusters being compared only, as opposed to all the clusters. This is why the input to the function is not a `WaveformPrincipalComponent` object but rather a `WaveformExtractor`. I think this raises bigger issues about the PC metrics (which you are probably aware of): I would imagine that in many cases (e.g. multielectrode array recordings) one would want to compute PCs of waveforms in only a subset of channels around the max channel, but currently that is not one of the `modes` in `compute_principal_components`. This potentially duplicates effort because one would then need to compute PCs for each cell's 'neighborhood'. I don't know how to get around this though. Thoughts?

Reply:
@khl02007 I understand. We could think about having a new method for pca called `neighborhood_concatenated` or `sparse_concatenated` which computes a concatenated model for each unit only on neighbor channels (which can be defined with the sparsity function). How does that sound? 

For now, however, let's keep this implementation as is. We can improve it later :)